### PR TITLE
Fix performance issue of syslogger_write_str()

### PR DESCRIPTION
**Pull request by yaowangm**
### Issue

In a real case CPU usage of logger process was extraordinarily high (more than 98%) and blocked other processes, 
causing seg down finally. Besides some improper settings, it's also due to a performance issue of logger code: In
`syslogger_write_str()`, `fwrite()`(or `write()`) was called per char instead of per string so it might be called for tens
of millions times when the log string is very big.

### Solution

We need to process the string by two kind of modifications (it might be the reason why existing code calls `fwrite()` 
per char instead of per string):

1. When terminate char ('\0') is found, finish the writing and ignore the rest.
2. When it is CSV format and '"' is found, write an additional '"' to implement the escape.

I was thinking about two solutions:

1. Use an extra buffer with double size (for the worst case), perform the modifications in buffer, and write the whole
 buffer. It can guarantee only one calling of `fwrite()`. However, the cost is extra memory requirement, and a lot of
memcopy in spite of how many '"' there are.
3. Go through the string, whenever hitting '"', write previous substring plus an extra '"', and continue. For the best case
(no '"' in the string), `fwrite()` is called only once. For the worst case (the entire string consist of '"'), the processing 
might fall back to "calling `fwrite()` per char", but no extra memory/memcopy is needed.

I adapted #2 since I suppose that log string containing a lot of '"' is not a common case.

### Test

I checked the log string in CSV manually. No additional test case is need because all operations writing log cover the scenario.

### CODE REVIEW

File: src/backend/postmaster/syslogger.c [hunk 1]
```diff
@@ -1188,29 +1198,56 @@ syslogger_append_current_timestamp(bool amsyslogger)
...
     if (data == NULL)
+	{
         return 0;
+	}
 
-    while (cnt < len && data[cnt] != '\0')
-    {
-        if (csv && data[cnt] == '"')
+	/*
+	 * We need to process the string by two kind of modifications:
```
Comment (ashwinstar) [inline src/backend/postmaster/syslogger.c:1210]: So all these seem GPDB specific functionality? Curious why we need to have diff here compared to upstream.

Comment (yaowangm) [reply src/backend/postmaster/syslogger.c:1210]: Yes. The function `syslogger_write_str()`, and almost the entire call stack is GPDB specific. See the stack:

> #0  0x00007f8b6b34473c in fwrite () from /data/logs/305770/packcore-core.392779/lib64/libc.so.6
> #1  0x0000000000a11029 in write_binary_to_file (fh=<optimized out>, count=1,
> buffer=0x2b6d1c0 "f','2022-03-11 00:17:46.94','2022-03-23 16:42:25.329497','MS_CUSIP','99C6Q26Z1'),('04938f3a-9574-4d68-9e24-d55842d3d72f','2022-03-11 00:17:46.94','2022-03-23 16:42:25.329497','APPLICATION','audit-hist"...) at syslogger.c:2176
> #2  write_syslogger_file_binary (count=1, destination=1,
> buffer=0x2b6d1c0 "f','2022-03-11 00:17:46.94','2022-03-23 16:42:25.329497','MS_CUSIP','99C6Q26Z1'),('04938f3a-9574-4d68-9e24-d55842d3d72f','2022-03-11 00:17:46.94','2022-03-23 16:42:25.329497','APPLICATION','audit-hist"...) at syslogger.c:2220
> #3  syslogger_write_str (data=<optimized out>, len=<optimized out>, amsyslogger=1 '\001', csv=<optimized out>) at syslogger.c:1313
> #4  0x0000000000a112ec in syslogger_write_str_with_comma (
> data=data@entry=0x2b6ce50 "execute <unnamed>: insert into audit_log.enriched_tags_data(id ,audit_ts , db_ts , namespace , tag) values('04938f3a-9574-4d68-9e24-d55842d3d72f','2022-03-11 00:17:46.94','2022-03-23 16:42:25.329493',"..., quote_empty=0 '\000',
> csv=1 '\001', amsyslogger=1 '\001') at syslogger.c:1524
> #5  0x0000000000a11ac0 in syslogger_log_chunk_list (chunk=chunk@entry=0x1aa1140 <saved_chunks+135168>) at syslogger.c:1898
> #6  0x0000000000a12c9c in syslogger_handle_chunk (chunk=0x1b4e140 <saved_chunks+843776>) at syslogger.c:2050
> #7  SysLoggerMain (argc=0, argv=0x0) at syslogger.c:689
> #8  SysLogger_Start () at syslogger.c:967
> #9  0x0000000000a0f585 in PostmasterMain (argc=argc@entry=6, argv=argv@entry=0x2b134f0) at postmaster.c:1427
> #10 0x00000000006b23f1 in main (argc=6, argv=0x2b134f0) at main.c:245

The function `SysLoggerMain()` exists in both postgresql and GPDB, but the chain from `syslogger_handle_chunk() `is GPDB specific. It seem GPDB implemented a totally different code path from the very top.

It's code to process pipe input in `SysLoggerMain()` of postgresql 12:

```
            bytesRead = read(syslogPipe[0],
                             logbuffer + bytes_in_logbuffer,
                             sizeof(logbuffer) - bytes_in_logbuffer);
            if (bytesRead < 0) 
            {    
 ...
            }    
            else if (bytesRead > 0) 
            {    
                bytes_in_logbuffer += bytesRead;
                process_pipe_input(logbuffer, &bytes_in_logbuffer);
                continue;
            }    
```

It's relevant code in GPDB (note the calling of `syslogger_handle_chunk()`):

```
            else 
            {    
                if (bytesRead + readPos >= sizeof(PipeProtoHeader) &&
                    chunk_is_postgres_chunk((PipeProtoHeader *)chunk))
                {    
                    int         chunk_size = chunk->hdr.len + sizeof(PipeProtoHeader);
                    int         needBytes = chunk_size - (bytesRead + readPos);

                    /*   
                     * Finish reading a chunk if the bytes we have read so far
                     * is not sufficient.
                     */
                    if (needBytes > 0) 
                    {    
                        bytesRead = read(syslogPipe[0],
                                             ((char *)chunk) + (bytesRead + readPos),
                                             needBytes);

                        Assert(bytesRead == needBytes);
                    }    

                    syslogger_handle_chunk(chunk);
```

**Review by soumyadeep2007 — COMMENTED** [review]
Replied offline. TL;DR using GPDB specific code we fixed the problem of multiplexed log lines in case of large 3rd party logs. This was later fixed in 2012 by Tom Lane (Commit: https://github.com/postgres/postgres/commit/c17e863bc7677a54d6da5bbb2868cca2cd9b30c1
Discussion: https://www.postgresql.org/message-id/flat/4F78D7FA.1070003%40dunslane.net)

We should remove the diff now.

**Comment by yaowangm** [comment]
Thanks @soumyadeep2007 and @ashwinstar. As discussed, I will remove the diff of `SysLoggerMain()` between postgresql and GPDB instead of fixing the performance issue. I will close the PR and open a new one when the code is ready.

**Pull request closed without merging**